File: 3332-minimum-operations-to-exceed-threshold-value-ii/minimum-operations-to-exceed-threshold-value-ii.cpp
#define ll long long
class Solution {
public:
    int minOperations(vector<int>& nums, int k) {
        int n=nums.size();
        int l=0,h=n;
        int ans=-1;
        auto f=[&](int mid)->bool{
            //cout<<mid<<endl;
            priority_queue<ll,vector<ll>,greater<ll>>pq;
            for(ll i:nums)pq.push(i);
            while(mid and pq.size()>=2){
            ll a=pq.top();
            pq.pop();
            ll b=pq.top();
            pq.pop();
            pq.push(min(a,b)*2+max(a,b));
            mid--;
            }
            while(!pq.empty()){
                if(pq.top()<k) return false;
                pq.pop();
            }
            return true;
        };
        while(l<=h){
            int mid=(l+h)>>1;
            if(f(mid)){
                ans=mid;
                h=mid-1;
            }
            else{
                l=mid+1;
            }
        }
        return ans;
    }
};

File: 3332-minimum-operations-to-exceed-threshold-value-ii/minimum-operations-to-exceed-threshold-value-ii.py
import heapq as hq

class Solution:
    def minOperations(self, nums: List[int], k: int) -> int:
        n=len(nums)
        pq=[]
        for i in nums:
            hq.heappush(pq,i)
        
        if(n<2):
            return pq[0]>=k
        
        cnt=0
        while len(pq)>=2:
            if pq[0]>=k:
                break
            a=hq.heappop(pq)
            b=hq.heappop(pq)
            hq.heappush(pq,min(a,b)*2+max(a,b))
            cnt+=1
            
        
        
        
        return cnt